configure.go

```diff
@@ -0,0 +1,146 @@
+	"bytes"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"os"
+	"reflect"
+	"strings"
+)
+
+const (
+	structTagKey             = "config"
+	requiredTagKey           = "required"
+	missingValuesErrTemplate = "Missing required fields: %s"
+)
+
+var (
+	ErrStringsOnly            = errors.New("Only string values are allowed in a config struct.")
+	ErrNotReference           = errors.New("The config struct must be a pointer.")
+	ErrStructOnly             = errors.New("Config object must be a struct.")
+	ErrNoTagValue             = errors.New("Config object attributes must have a 'config' tag value.")
+	ErrTooManyTagValues       = errors.New("Config object attributes can only have a key and optional required attribute.")
+	ErrFlagParsed             = errors.New("The flag library cannot be used in conjunction with configure")
+	ErrInvalidJSON            = errors.New("Invalid JSON found in arguments.")
+	ErrStructTagInvalidOption = errors.New("Only 'required' is a config option.")
+)
+
+func parseTagKey(tag string) (key string, required bool, err error) {
+	if tag == "" {
+		return "", false, ErrNoTagValue
+	}
+
+	s := strings.Split(tag, ",")
+	switch len(s) {
+	case 2:
+		if s[1] != requiredTagKey {
+			return "", false, ErrStructTagInvalidOption
+		}
+		return s[0], true, nil
+	case 1:
+		return s[0], false, nil
+	default:
+		return "", false, ErrTooManyTagValues
+	}
+}
+
+// Configure takes a reference to an interface that has 'config' tags on all atttributes of
+// the struct. Configure first tries to find values for these attributes through command line
+// flags, then will attempt to parse the first argument as a JSON blob.
+// An attribute can be required by appending ',required' to the config key.
+func Configure(configStruct interface{}) error {
+	if flag.Parsed() {
+		return ErrFlagParsed
+	}
+
+	reflectConfig := reflect.ValueOf(configStruct)
+	if reflectConfig.Kind() != reflect.Ptr {
+		return ErrStructOnly
+	}
+
+	var (
+		configFlags  = flag.NewFlagSet("configure", flag.ContinueOnError)
+		flagValueMap = map[string]*string{} // holds references to attribute flags
+		flagFound    = false                // notes if any flags are found, JSON parsing is skipped if so
+	)
+
+	config := reflectConfig.Elem()
+	for i := 0; i < config.NumField(); i++ {
+		valueField := config.Field(i)
+		if !valueField.CanSet() {
+			return ErrNotReference
+		}
+
+		typedAttr := config.Type().Field(i)
+		if typedAttr.Type.Name() != "string" {
+			return ErrStringsOnly
+		}
+
+		tagVal, _, err := parseTagKey(typedAttr.Tag.Get(structTagKey))
+		if err != nil {
+			return err
+		}
+
+		flagValueMap[tagVal] = configFlags.String(tagVal, "", "generated field")
+	}
+	if err := configFlags.Parse(os.Args[1:]); err != nil {
+		return err
+	}
+
+	for i := 0; i < config.NumField(); i++ {
+		valueField := config.Field(i)
+		tagVal, _, err := parseTagKey(config.Type().Field(i).Tag.Get(structTagKey))
+		if err != nil {
+			return err
+		}
+
+		if *flagValueMap[tagVal] != "" {
```

Comment:
I having trouble understanding what the code I was reading did, I think because a lot of the names are similar. Can you maybe put a comment at the beginning of each of for loops describing what it does? e.g. "parse tags from struct and create flagset based on them"